source/common/upstream/upstream_impl.cc

```diff
@@ -909,6 +912,9 @@ bool BaseDynamicClusterImpl::updateDynamicHostList(const HostVector& new_hosts,
         i = current_hosts.erase(i);
         found = true;
       } else {
+        if (address_matched && health_check_changed) {
```

Comment:
Sorry little confused, why is this change necessary? Same below? Seems like just pivoting on address + HC address should be enough above?

Reply:
Tried by only introducing the change above. But from my test, without this, it leads to keeping the old hosts.

Reply:
OK, can you take a look at why that is? This code is getting really complicated and could use some more comments anyway. :)

Reply:
Sure. Will try to trace out the flow and craft some comments.

Reply:
Referring to the case example by @bplotnick in here: https://github.com/bplotnick/envoy_examples/blob/master/healthcheck_updates/envoy_config/clusters.yaml, the `drain_connections_on_host_removal` is not being set (as true) for the EDS cluster config (`type: EDS`). If this set to true, we don't need these two changes. Do you think if we have health checker, we should always set this flag to true?

The flag is introduced by @snowp in https://github.com/envoyproxy/envoy/commit/08712e93b07695f53d192a2601cfa2ccc7a20f33

Reply:
These two changes make sure, regardless the flag, we always rebuild and skip keeping healthy (old/current) hosts.

Reply:
Heh whoops. That was actually unintentional. In my original configuration where I saw this issue, I had that flag set 😛

Reply:
But probably that the correct behavior? Once the old hosts die, it will be removed eventually.

Reply:
I need a little time to grok this conversation thread. I will follow up today or tomorrow. Thank you!